### PR TITLE
[DataFrame] Implement diff

### PR DESCRIPTION
**Pull request by 11rohans**
Implements `DataFrame.diff`

### CODE REVIEW

**Comment by kunalgosar** [comment]
Dtypes PR was merged in so this can be rebased on master now.

File: python/ray/dataframe/dataframe.py [hunk 1]
```diff
@@ -1291,9 +1331,30 @@ def describe_helper(df):
...
+
+        Args:
+            periods: Periods to shift for forming difference
+            axis: Take difference over rows or columns
+
+        Returns:
+            DataFrame with the diff applied
+        """
+
+        if (axis == 1 or axis == 'columns'):
```
Comment (kunalgosar) [inline python/ray/dataframe/dataframe.py]: Can use `axis = pd.DataFrame()._get_axis_number(axis)` here.

File: python/ray/dataframe/dataframe.py [hunk 2]
```diff
@@ -1291,9 +1331,30 @@ def describe_helper(df):
...
+            result = _map_partitions(lambda df:
```
Comment (kunalgosar) [inline python/ray/dataframe/dataframe.py]: Some code can be reused between the two functions here. There are examples that do similar things, eg: `mode`.

File: python/ray/dataframe/test/test_dataframe.py [hunk 3]
```diff
@@ -1141,11 +1144,9 @@ def test_describe(ray_df, pandas_df):
     assert(ray_df.describe().equals(pandas_df.describe()))
 
 
-def test_diff():
-    ray_df = create_test_dataframe()
-
-    with pytest.raises(NotImplementedError):
-        ray_df.diff()
+@pytest.fixture
+def test_diff(ray_df, pandas_df):
```
Comment (kunalgosar) [inline python/ray/dataframe/test/test_dataframe.py]: Try different axis and period values in the tests.

**Review by kunalgosar — CHANGES_REQUESTED** [review]
Looks really good! Probably doesn't require the rebase on #1930 and #1956 though.

**Comment by 11rohans** [comment]
Comments resolved

**Review by kunalgosar — APPROVED** [review]
Looks great!

**Comment by kunalgosar** [comment]
Can you resolve flake issues?

**Comment by 11rohans** [comment]
resolved flake8

**Review by devin-petersohn — APPROVED** [review]

**Comment by devin-petersohn** [comment]
Passes private-travis. Merged, thanks @11rohans

**Pull request merged**